consensus/view_change_construct.go

```diff
@@ -0,0 +1,422 @@
+		}
+		_, ok := vc.bhpSigs[recvMsg.ViewID][senderKeyStr]
+		if ok {
+			return M1, nil, errors.New("[VerifyViewChangeMsg] received M1 (prepared) message already")
+		}
+		if !recvMsg.ViewchangeSig.VerifyHash(senderKey.Object, recvMsg.Payload) {
+			return ERR, nil, errors.New("[VerifyViewChangeMsg] failed to verify signature for m1 type message")
+		}
+		blockHash := recvMsg.Payload[:32]
+
+		aggSig, mask, err := chain.ReadSignatureBitmapByPublicKeys(recvMsg.Payload[32:], members)
+		if err != nil {
+			return ERR, nil, err
+		}
+		if !aggSig.VerifyHash(mask.AggregatePublic, blockHash[:]) {
+			return ERR, nil, errors.New("[VerifyViewChangeMsg] failed to verify multi signature for m1 prepared payload")
+		}
+
+		vc.getLogger().Info().Uint64("viewID", recvMsg.ViewID).
+			Str("validatorPubKey", senderKeyStr).
+			Msg("[VerifyViewChangeMsg] Add M1 (prepared) type message")
+		vc.bhpSigs[recvMsg.ViewID][senderKeyStr] = recvMsg.ViewchangeSig
+		vc.bhpBitmap[recvMsg.ViewID].SetKey(senderKey.Bytes, true) // Set the bitmap indicating that this validator signed.
+
+		vc.getLogger().Info().Uint64("viewID", recvMsg.ViewID).
+			Str("validatorPubKey", senderKeyStr).
+			Msg("[VerifyViewChangeMsg] Add M3 (ViewID) type message")
+
+		vc.viewIDSigs[recvMsg.ViewID][senderKeyStr] = recvMsg.ViewidSig
+		// Set the bitmap indicating that this validator signed.
+		vc.viewIDBitmap[recvMsg.ViewID].SetKey(senderKey.Bytes, true)
+
+		return M1, preparedBlock, nil
+	}
+	_, ok := vc.nilSigs[recvMsg.ViewID][senderKeyStr]
+	if ok {
+		return M2, nil, errors.New("[VerifyViewChangeMsg] received M2 (NIL) message already")
+	}
+	if !recvMsg.ViewchangeSig.VerifyHash(senderKey.Object, NIL) {
+		return ERR, nil, errors.New("[VerifyViewChangeMsg] failed to verify signature for m2 type message")
+	}
+
+	vc.getLogger().Info().Uint64("viewID", recvMsg.ViewID).
+		Str("validatorPubKey", senderKeyStr).
+		Msg("[VerifyViewChangeMsg] Add M2 (NIL) type message")
+	vc.nilSigs[recvMsg.ViewID][senderKeyStr] = recvMsg.ViewchangeSig
+	vc.nilBitmap[recvMsg.ViewID].SetKey(recvMsg.SenderPubkey.Bytes, true) // Set the bitmap indicating that this validator signed.
+
+	vc.getLogger().Info().Uint64("viewID", recvMsg.ViewID).
+		Str("validatorPubKey", senderKeyStr).
+		Msg("[VerifyViewChangeMsg] Add M3 (ViewID) type message")
+
+	vc.viewIDSigs[recvMsg.ViewID][senderKeyStr] = recvMsg.ViewidSig
+	// Set the bitmap indicating that this validator signed.
+	vc.viewIDBitmap[recvMsg.ViewID].SetKey(senderKey.Bytes, true)
+
+	return M2, nil, nil
+}
+
+// InitPayload do it once when validator received view change message
+// TODO: if I start the view change message, I should init my payload already
+func (vc *viewChange) InitPayload(
+	fbftlog *FBFTLog,
+	viewID uint64,
+	blockNum uint64,
+	myPubKey string,
+	privKeys multibls.PrivateKeys,
+) error {
+	vc.vcLock.Lock()
+	defer vc.vcLock.Unlock()
+
+	// add my own M2 type message signature and bitmap
+	_, ok := vc.nilSigs[viewID][myPubKey]
+
+	if !ok {
```

Comment:
The logic should be as the original condition `if !(ok1 || ok2) {`  . meaning: when I never added either m1 or m2 myself, I proceed to check whether I have prepared block myself, if yes, I add to m1, otherwise, I add to m2. Honest node can only commit to one choice, either m1 or m2. They are mutual exclusive. Let's simply keep the original logic (copy and paste is better to avoid mistakes)

Reply:
okay,